core/src/main/java/io/questdb/cairo/TableWriter.java

```diff
@@ -4361,28 +4367,131 @@ private void o3CommitPartitionAsync(
-
-                o3PartitionUpdate(
-                        timestampMin,
-                        partitionTimestamp,
-                        newPartitionSize,
-                        oldPartitionSize,
-                        partitionMutates,
-                        isLastWrittenPartition
-                );
+                final long srcDataNewPartitionSize = Unsafe.getUnsafe().getLong(blockAddress + 2 * Long.BYTES);
+                final long srcDataOldPartitionSize = Unsafe.getUnsafe().getLong(blockAddress + 3 * Long.BYTES);
+                final long flags = Unsafe.getUnsafe().getLong(blockAddress + 4 * Long.BYTES);
+                final boolean partitionMutates = Numbers.decodeLowInt(flags) != 0;
+                final boolean isLastWrittenPartition = o3PartitionUpdateSink.nextBlockIndex(blockIndex) == -1;
+                final long o3SplitPartitionSize = Unsafe.getUnsafe().getLong(blockAddress + 5 * Long.BYTES);
+
+                txWriter.minTimestamp = Math.min(timestampMin, txWriter.minTimestamp);
+                int partitionIndexRaw = txWriter.findAttachedPartitionRawIndexByLoTimestamp(partitionTimestamp);
+
+                final long newPartitionTimestamp = partitionTimestamp;
+                final int newPartitionIndex = partitionIndexRaw;
+                if (partitionIndexRaw < 0) {
+                    // This is partition split. Instead of rewriting partition because of O3 merge
+                    // the partition is kept, and its tail rewritten.
+                    // The new partition overlaps in time with the previous one.
+                    partitionTimestamp = txWriter.getPartitionTimestampByTimestamp(partitionTimestamp);
+                    partitionIndexRaw = txWriter.findAttachedPartitionRawIndexByLoTimestamp(partitionTimestamp);
+                }
+
+                if (partitionTimestamp == lastPartitionTimestamp && newPartitionTimestamp == partitionTimestamp) {
+                    if (partitionMutates) {
+                        // Last partition is rewritten.
+                        closeActivePartition(true);
+                    } else if (!isLastWrittenPartition) {
+                        // Last partition is appended, and it is not the last partition anymore.
+                        closeActivePartition(srcDataNewPartitionSize);
+                    } else {
+                        // Last partition is appended, and it is still the last partition.
+                        setAppendPosition(srcDataNewPartitionSize, false);
+                    }
+                }
+
+                if (partitionTimestamp < lastPartitionTimestamp) {
+                    // increment fixedRowCount by number of rows old partition incremented
+                    this.txWriter.fixedRowCount += srcDataNewPartitionSize - srcDataOldPartitionSize + o3SplitPartitionSize;
+                } else if (partitionTimestamp == lastPartitionTimestamp) {
```

Comment:
nit: I think this would be more readable to collapse 2 else into 1

```suggestion
                } else {
                    if (partitionTimestamp > lastPartitionTimestamp) {
                         // this is potentially a new last partition
                         this.txWriter.fixedRowCount += commitTransientRowCount;
                    }
                     
                      // was this "new" partition split?
                      if (o3SplitPartitionSize > 0) {
                          // yep, it was
                          // the "current" partition instantly becomes fixed
                          this.txWriter.fixedRowCount += srcDataNewPartitionSize;
                          commitTransientRowCount = o3SplitPartitionSize;
                      } else {
                          commitTransientRowCount = srcDataNewPartitionSize;
                      } 
                    }
```

Reply:
done in https://github.com/questdb/questdb/pull/3704/commits/1524b68c48b81fb28f121b37fd84a402285ade43